### PR TITLE
Restore NIX_PATH taking precedence over nix-path

### LINKED ISSUE #7857: NIX_PATH no longer overrides nix-path
**Describe the bug**

In Nix 2.13, the `NIX_PATH` environment variable overrides the value of the `nix-path` setting in `nix.conf`. 

**Steps To Reproduce**

E.g. if `nix.conf` contains
```
extra-nix-path = nixpkgs=flake:nixpkgs
```
then Nix 2.13 behaves as follows:
```
# NIX_PATH=nixpkgs=flake:foo nix-build  '<nixpkgs>' -A hello
error: cannot find flake 'flake:foo' in the flake registries
```

But on current master, `NIX_PATH` is ignored:

```
# NIX_PATH=nixpkgs=flake:foo nix-build  '<nixpkgs>' -A hello
/nix/store/7j2sjbdbhlyda1sm0s6p0frfy0dxj67i-hello-2.12.1
```

**Expected behavior**

`NIX_PATH` should take precedence. Not only is that what most people would expect (I think), but it's also a regression.

**`nix-env --version` output**

**Additional context**

This is probably caused by https://github.com/NixOS/nix/commit/1cba5984a68a489c4a56691032e4c87991c678f4.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### LINKED ISSUE #8784: `nix-env --query --available` does not follow `NIX_PATH`
**Describe the bug**

My system does not have any channels.

I tried to use `nix-index` today, but it failed.

```
$ nix-index
+ querying available packages
error: querying available packages failed
caused by: nix-env failed with error: nix-env failed with exit code 1:
error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)

       at «none»:0: (source not available)

```

Taking the `nix-env` hint, I tried the following:

```
$ cd nixpkgs

$ NIX_PATH=nixpkgs=$PWD nix-env -qaP

```

ie no output

`strace` reveals that my nixpkgs checkout wasn't accessed at all.

<details>

```
$ NIX_PATH=nixpkgs=$PWD strace nix-env -qaP 2>&1 | grep /home
newfstatat(AT_FDCWD, "/home/user", {st_mode=S_IFDIR|0700, st_size=6286, ...}, 0) = 0
newfstatat(AT_FDCWD, "/home/user/.nix-defexpr/channels", {st_mode=S_IFLNK|0777, st_size=44, ...}, AT_SYMLINK_NOFOLLOW) = 0
openat(AT_FDCWD, "/home/user/.nix-profile/etc/xdg/nix/nix.conf", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/user/.config/nix/nix.conf", O_RDONLY|O_CLOEXEC) = 3
newfstatat(AT_FDCWD, "/home/user/.nix-defexpr", {st_mode=S_IFDIR|0755, st_size=42, ...}, AT_SYMLINK_NOFOLLOW) = 0
newfstatat(AT_FDCWD, "/home/user/.nix-profile", {st_mode=S_IFLNK|0777, st_size=43, ...}, AT_SYMLINK_NOFOLLOW) = 0
readlink("/home/user/.nix-profile", "/nix/var/nix/profiles/per-user/u"..., 1024) = 43
newfstatat(AT_FDCWD, "/home/user/.nix-defexpr", {st_mode=S_IFDIR|0755, st_size=42, ...}, 0) = 0
newfstatat(AT_FDCWD, "/home/user/.nix-defexpr/default.nix", 0x7ffd0c005550, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/user/.nix-defexpr", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 3
newfstatat(AT_FDCWD, "/home/user/.nix-defexpr/channels", 0x7ffd0c005590, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/home/user/.nix-defexpr/channels_root", 0x7ffd0c005590, 0) = -1 ENOENT (No such file or directory)

```

</details>

I've tried with a [nix-defexpr](https://nixos.org/manual/nix/unstable/command-ref/nix-env/query#options)

```
$ cat ~/.nix-defexpr/default.nix 
import <nixpkgs> {}

$ NIX_PATH=nixpkgs=$PWD nix-env -qaP
error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)

       at /home/user/.nix-defexpr/default.nix:1:8:

            1| import <nixpkgs> {}
             |        ^
            2|
(use '--show-trace' to show detailed location information)

```

**Steps To Reproduce**

1. (optionally?) Configure NixOS with `nix.channel.enable = false;`
2. Remove all channels
3. Run `nix-env` with a `NIX_PATH` or `-I`.

**Expected behavior**

I can use `nix-env` and `nix-index` with just `NIX_PATH` and no channels on my system.

**`nix-env --version` output**

**Additional context**

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### PR DESCRIPTION
**Pull request by edolstra**
# Motivation

This restores the Nix <= 2.13 behaviour. Environment variables (which are dynamic) should take precedence over configuration file settings (which are static).

- Fixes #7857.
- Fixes #8902 (duplicate?)
- Fixes #8784

Note however that it does make `NIX_PATH` take precedence over the `--nix-path` and `--extra-nix-path` command line options, due to the way settings are implemented.  There really is no correct behaviour here, e.g. should `--extra-nix-path` override `NIX_PATH` or the configuration file? If `--extra-nix-path` is used in restricted eval mode, should it extend the default value (which includes the channels)?

Bottom line: it's best to stop using restricted eval and/or `NIX_PATH`.

# Context

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

### CODE REVIEW

**Comment by edolstra** [comment]
Alternatively, we can revert #7689, which has the benefit of simply restoring the old behaviour.

**Comment by fricklerhandwerk** [comment]
I think there is correct behavior, and this is strictly

> settings < environment variables < command line options

If the current code structure doesn't allow expressing it cleanly, probably that structure should be changed. I understand that this would be quite some work for mostly a side issue, and I think that the change in this PR is a net improvement over both the current state and reverting #7689.

File: src/libexpr/eval.cc [hunk 1]
```diff
@@ -519,33 +519,40 @@ EvalState::EvalState(
     static_assert(sizeof(Env) <= 16, "environment must be <= 16 bytes");
 
     /* Initialise the Nix expression search path. */
-    evalSettings.nixPath.setDefault(evalSettings.getDefaultNixPath());
     if (!evalSettings.pureEval) {
```
Comment (fricklerhandwerk) [inline src/libexpr/eval.cc:522]: Is there anything speaking against

```cpp
if (evalSettings.pureEval) {
  createBaseEnv();
  return;
}
```
and then increasing looseness in the following?

The inverted logic makes reasoning quite cumbersome.

**Comment by grahamc** [comment]
> 
> I would say the order would be this:
> 
>     1. Configuration file
> 
>     2. `NIX_PATH` overrides configuration file
> 
>     3. `--nix-path` overrides `NIX_PATH`
> 
>     4. `--extra-nix-path` extends current value, whatever it may be.
> 

That makes good sense to me, personally.

**Comment by edolstra** [comment]
There are some fundamentally conflicting requirements here. `--restrict-eval` is supposed to ignore the `nix-path` setting, but not `NIX_PATH`. (Maybe that was a mistake...) So we can't implement `NIX_PATH` as just another way to override `nix-path`. It has to have special treatment, where it takes precedence over `nix-path`. But since `--[extra-]nix-path` is a way to override/extend the `nix-path` setting, it has no effect if `NIX_PATH` is set. BTW, there really is no need to ever use `--extra-nix-path` since `-I` exists which does work as expected.

Anyway, I've made #7911 which has the simpler fix of reverting #7689. I'll merge that before the next release tomorrow.

**Comment by fricklerhandwerk** [comment]
> Bottom line: it's best to stop using [...] `NIX_PATH`.

This, very much. `NIX_PATH` was clearly a mistake that opened the floodgates for impurity, and impurity is the root of all confusion.

But getting away from `NIX_PATH` and channels is mostly a social problem that can be helped with better documentation and lots of convincing people to spend time and migrate their code to other ways of doing things.

**Comment by Ericson2314** [comment]
> (Maybe that was a mistake...)

Yeah I think it is. IMO `--restrict-eval` is kinda ad-hoc (as opposed to the more principled pure eval) and I think it would be good to change this so the ordering is consistent with everything else.

**Comment by fricklerhandwerk** [comment]
Documented the desired behavior in https://github.com/NixOS/nix/pull/7772. One could merge that and treat deviating behavior as a bug to be fixed.

**Comment by fricklerhandwerk** [comment]
Discussed in the Nix team meeting:


- @edolstra: the configuration system is currently built in a way where the proposed behavior cannot really be expressed. options:
    - we could put a lot of effort into this
    - ignore the problem
    - deprecate `restrict-eval`
        - Hydra uses it, can't just rip it out
- @ericson2314: in favor of deprecating `restrict-eval` as long as Hydra can get this configuration somehow
    - may need a proper policy for deprecation an breaking changes, e.g. as proposed by @infinisil
- @roberth: Hydra doesn't use the CLI. We could deprecate the CLI but keep the mechanism internally.
- @edolstra: currently working on filesystem accessor from lazy trees; could use that to implement a policy in terms of a filter according to Hydra's needs.
- @thufschmitt: can add a warning with a prompt to complain in a GitHub issue to figure out how much the feature is used
- @fricklerhandwerk: originally the underlying issue was that `restrict-eval` was is not respected. the PR fixing that exposed an inconsistency in the CLI parameter handling
    - @edolstra: the problem is that there are a few ad-hoc legacy mechanisms interacting in weird ways, especially `$NIX_PATH`.
- @thufschmitt: `restrict-eval` not working exactly is less of a problem than `NIX_PATH` not working, we can postpone it
- @fricklerhandwerk will open an issue for adding a deprecation warning to `restrict-eval`

**Comment by roberth** [comment]
I'm having trouble with this sometimes because I use `nix-path = ` (empty) through NixOS via `nix.channel.enable = false;`. Anyone who opts into that won't be able to use `NIX_PATH` temporarily when they need it for whatever mediocre reason (such as a nix-shell that makes `<nixpkgs>` more hermetic).

I think we [reached](https://github.com/NixOS/nix/pull/7871#issuecomment-1536134468) a decision. Is this blocked on anything?

**Comment by katandtonic** [comment]
Please fix this!

**Comment by fricklerhandwerk** [comment]
Superseded by https://github.com/NixOS/nix/pull/11079
